### ISSUE TITLE
Keep aliases when transforming plugin JSON to Style Dictionary JSON

### ISSUE THREAD

**Issue by jakobe** [six7/figma-tokens#305]
**Is your feature request related to a problem? Please describe.**
First of all love what you've created!

I'm using [Token Transformer as described in the docs](https://docs.tokens.studio/sync/github#7-how-to-use-tokens-stored-in-github-in-development)

It does state that:
> You can then use it to generate tokens to a JSON that is readable by Style Dictionary, without any aliases, without any math.

While this is great for the math part (e.g. `{baseUnit} * 2`) as SD doesn't have support for that, I would like it to keep the regular aliases (e.g. `{color.primary}`) in the output to allow for SD's [Output references](https://amzn.github.io/style-dictionary/#/version_3?id=output-references) introduced in v3.

**Describe the solution you'd like**
This could either be added as CLI option or - even better - just changed to keep the aliases in the output, as SD already have support for that and can either be translated to actual values (SD default) or configured to be outputted as references.

**Describe alternatives you've considered**
🤷🏻 

**Additional context**

**Comment by six7** [comment]
I agree. I think what would be good is for token-transformer to decide on whether the value is style-dictionary conform or is using custom Figma Tokens features such as Math and `rgba($color, 0.5)`. We should be able to do that quite easily by just checking if the value only contains an alias and nothing else.

Might even be able to add `.value` at the end there?

**Comment by jakobe** [comment]
> I agree. I think what would be good is for token-transformer to decide on whether the value is style-dictionary conform or is using custom Figma Tokens features such as Math and `rgba($color, 0.5)`. We should be able to do that quite easily by just checking if the value only contains an alias and nothing else.
> 
> Might even be able to add `.value` at the end there?

Great :)

Yes excactly - I'm currently "patching"/escaping the original aliases coming from the plugin export, by changing `{alias}` to `/alias/` using a regex to search for alias values: `/"{((\w+\.?)+)}"/gm` to avoid token-transformer resolving them. Then after the transformation I'm reverting them back from `/alias/` to `{alias}` and at the same time suffixing the alias with `.value`, i.e. `{alias}` to `{alias.value}` to allow SD to do the transformation based on the config.

But it would be really great if that was already baked into token-transformer. Do you want me to create a PR or do you have time to do it? I had a brief look at the implementation of token-transformer, and I can see it uses transformations from the plugin, so not really sure if that's the right place to change it as it might break the plugin in itself?
Let me know if you want me to look at it, and maybe point me in the right direction 🤞🏻

**Comment by jakobe** [comment]
@six7 Just adding some additional info to this:
As stated in https://github.com/six7/figma-tokens/issues/284#issuecomment-1006352476 Style Dictionary has made it it optional to suffix references/aliases with `.value` (e.g. `{color.blue.value}` vs just `{color.blue}`) which also aligns with how aliases/references are defined in the [Design Tokens Format Module](https://design-tokens.github.io/community-group/format/#aliases-references).

So I guess:

> Might even be able to add .value at the end there?

isn't relevant anymore.

**Comment by six7** [comment]
I think it'd be great if you could work on this! @jeffreydelooff recently introduced a better way of how we can do CLI parameters including an option to `expandTypography`. I'd like it if we introduced an option to `resolveReferences` - which by default could be set to true as it is currently.

The transformer uses parts of the plugin logic to not have to build things twice. We're then exporting a subset of functions to be used for token-transformer, the entry point for that is `src\utils\transform.ts` which is using `src\utils\convertTokensObjectToResolved.ts` where we basically also resolve all token values. I guess if the `resolveReferences `option is false then we'd not need that step.

**Comment by jakobe** [comment]
Hey @six7 ,

I finally got around to have a look at this.
And thx for the pointers, that helped me a lot 👍🏻 

Regarding introducing a cli option (i.e.`resolveReferences`): that's pretty clear with the work @jeffreydelooff already has done 👌🏻 

Regarding keeping aliases: This might be a bit trickier depending on the solution we choose...

We could skip the resolve step entirely as [you mentioned in your last comment ](https://github.com/six7/figma-tokens/issues/305#issuecomment-1019454836) but that would also skip resolving any [Figma Tokens plugin compliant math expressions](https://docs.tokens.studio/tokens/using-math), so not sure we wanna do that? Especially since SD currently doesn't support that.

[As we talked about earlier](https://github.com/six7/figma-tokens/issues/305#issuecomment-962893442) maybe we should instead distinguish between SD compliant aliases and Figma Tokens Math expressions?
E.g. SD: `{color.primary}` or `{brand}.{primary}`
vs Figma Tokens plugin: `1 * {sizing.base}` or `{sizing.small} * {sizing.scale}`

The question is whether we then whitelist math expressions (e.g. `+`, `-`, `*`, `/`) specific to the plugin when checking the original/raw value or we check for SD compliant aliases? Which could be infinite as both `{color.primary}`, `{brand}.{primary}` and `{basesize}rem` are all valid SD expressions, so I would prefer checking specifically for plugin math expressions instead.

We could also go with skipping the resolve step entirely which then just means that one shouldn't use any expression (e.g. math) that you don't intend to evaluate yourself later in the processing pipeline 🤷🏻 

I've been POC'ing locally on this, but haven't created a fork/PR yet as the implementation / affected files and function depends very in much on the above, but I'll be happy to do that right after we settle on the right solution 👍🏻

**Comment by jakobe** [comment]
@six7 Would you have time to chip in on this ☝🏻 🙏🏻

**Comment by six7** [comment]
> @six7 Would you have time to chip in on this ☝🏻 🙏🏻

Sorry, that must have slipped through my notifications 🙏

What if resolveReferences had multiple options: `false`, `true`, `math`:

`True`: Current
`False`: We don't resolve any references
`Math`: We only resolve references that contain math?

I think just checking for these expressions won't be enough as there's more than just these - e.g. `roundTo()`. What if we made the math evaluation optional? We could compare string before / after - if it's changed it would mean that math was present and we should use this one (if resolveReferences is set to math?

The evaluation happens in `src/utils/aliases.tsx` in `checkAndEvaluateMath`

I was also working a bit on this, maybe this PR could prove a good starting point for you? This one only has false/true - but I think you're right that we should incorporate Math into this!

**Comment by jakobe** [comment]
Thx @six7 good points 👍🏻 
I'll have a look later this week 🤞🏻

**Comment by nan4k7** [comment]
Any news on this? I'm having the same problem with aliases.

Thanks @six7 for your great work!

**Comment by jakobe** [comment]
> Any news on this? I'm having the same problem with aliases.
> 
> Thanks @six7 for your great work!

Sorry @nan4k7  unfortunately I haven't had time to look at this yet. Hope to find time within the next couple of weeks 🤞🏻

**Comment by mkitzmann** [comment]
We are currently having the same issue and we are not using any math expressions. 
Could PR #495 already be merged and later on be extended to ignore math expression? 
As far as I understood, `resolveReferences` would be `true` by default, so this PR should not have any breaking changes and also make it possible to extend it later on.

Thank you @jakobe and @six7 for your work!

**Comment by jakobe** [comment]
> We are currently having the same issue and we are not using any math expressions. Could PR #495 already be merged and later on be extended to ignore math expression? As far as I understood, `resolveReferences` would be `true` by default, so this PR should not have any breaking changes and also make it possible to extend it later on.
> 
> Thank you @jakobe and @six7 for your work!

Hi @mkitzmann and @nan4k7, I finally got around to do a PR on this 😅 
Regarding PR #495 that was done before a major refactor, so it cannot be easily merged.

But anyhoo, I've done a PR as described in this issue ☝🏻 , which adds the `resolveReferences` option to Token Transformer:
true: Current, resolves all aliases and math expressions
false: We don't resolve any references or math expressions
'math': We resolve references that contain math or concatened tokens (e.g. `{text.base}{text.unit}`)

@six7 I solved it a bit differently from what we discussed: It occurred to me that SD only supports single alias (as in `{sizing.base}`) when using the `outputReferences` option, but not concatenation (e.g. `{text.base}{text.unit}`or `{text.base}rem`) as the token needs to be translated into a variable on its own. Hope it makes sense 🤞🏻 

Let men know when you have time to review this and if anything is missing. I've added 3 additional tests to ensure everything works as described. 

Cheers,
Jakob 😄
